packages/clay-data-provider/src/types.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import LRU from 'lru-cache';
+
+export const SYMBOL_DATA_PROVIDER = Symbol('clay.data.provider');
+
+export const SYMBOL_ORIGIN = Symbol('clay.internal');
+
+/**
+ * - Loading (1) The status is set to `loading` only when the first
+ *   requisition occurs.
+ * - Refetch (2) The status is set to `refetch` when a change in the variables API
+ *   or refetch method is called.
+ * - Polling (3) The status is set to `polling` when pollInterval is set above 0.
+ * - Unused (4) When no request is happening the status will be `unused`.
+ * - Error (5) When any timeout or request `error` occurs, the status will be set
+ *   to error.
+ */
+export enum NetworkStatus {
+	Error = 5,
+	Loading = 1,
+	Polling = 3,
+	Refetch = 2,
+	Unused = 4,
+}
+
+export enum FetchPolicy {
+	CacheFirst = 'cache-first',
+	NoCache = 'no-cache',
+	CacheAndNetwork = 'cache-and-network',
+}
+
+export type TSymbolData = LRU<{}, {}>;
+
+export type TLink = Function | string;
+
+export type TVariables = Object | null;
+
+export interface IStorage {
+	[SYMBOL_DATA_PROVIDER]?: TSymbolData;
+	[SYMBOL_ORIGIN]?: boolean;
+	[propName: string]: any;
+}
+
+export interface IFetchRetryDelay {
+	/**
+	 * The number of milliseconds to wait before attempting the first retry.
+	 *
+	 * Delays will increase exponentially for each attempt.  E.g. if this is
+	 * set to 100, subsequent retries will be delayed by 200, 400, 800, etc...
+	 *
+	 * Note that if jittering is enabled, this is the average delay.
+	 */
+	initial?: number;
+
+	/**
+	 * Whether delays between attempts should be randomized.
+	 *
+	 * This helps avoid thundering herd type situations by better
+	 * distributing load during major outages.
+	 */
+	jitter?: boolean;
+}
+
+export interface IFetchRetry {
+	/**
+	 * The maximum number of times to try a single request before giving up.
+	 */
+	attempts?: number;
+
+	/**
+	 * Configuration for the delay strategy to use.
+	 */
+	delay?: IFetchRetryDelay;
+}
+
+export interface IDataProvider {
+	/**
+	 * This API is used in conjunction with variables API, if it is always
+	 * changing its value set a debounce time to make a new request.
+	 *
+	 * Set a value in ms.
+	 */
+	fetchDelay?: number;
+
+	/**
+	 * Set ups the request options.
```

Comment:
"Set ups" doesn't make much sense here. Maybe we could do "Options passed to request configuration"